src/lib/helpers.ts

```diff
@@ -50,3 +50,69 @@ export async function createDirectoryIfNotExists(path: string) {
     }
   }
 }
+
+export function hasEveryOverlap<T>(array1: T[], array2: T[]): boolean {
+  if (!Array.isArray(array1) || !Array.isArray(array2)) {
+    return false;
+  }
+
+  return array1.every(element => array2.includes(element));
+}
+
+export function hasAnyOverlap<T>(array1: T[], array2: T[]): boolean {
+  if (!Array.isArray(array1) || !Array.isArray(array2)) {
+    return false;
+  }
+
+  return array1.some(element => array2.includes(element));
+}
+
+export function ignoredNamespaceConflict(ignoreNamespaces: string[], bindingNamespaces: string[]) {
+  return hasAnyOverlap(bindingNamespaces, ignoreNamespaces);
+}
+
+export function bindingAndCapabilityNSConflict(bindingNamespaces: string[], capabilityNamespaces: string[]) {
+  if (!capabilityNamespaces) {
+    return false;
+  }
+  return capabilityNamespaces.length !== 0 && !hasEveryOverlap(bindingNamespaces, capabilityNamespaces);
+}
+
+export function generateWatchNamespaceError(
+  ignoredNamespaces: string[],
+  bindingNamespaces: string[],
+  capabilityNamespaces: string[],
+) {
+  let err = "";
+
+  // check if binding uses an ignored namespace
+  if (ignoredNamespaceConflict(ignoredNamespaces, bindingNamespaces)) {
+    err += "Binding uses a Pepr ignored namespace.";
+  }
+
+  // ensure filter namespaces are part of capability namespaces
+  if (bindingAndCapabilityNSConflict(bindingNamespaces, capabilityNamespaces)) {
+    err += "Binding uses namespace not governed by capability.";
+  }
```

Comment:
I was kinda surprised that these error messages don't announce _which namespace/s_ is/are causing the error..?  Is there a reason we're not just outputting what's in conflict?  Seems kinda awkward to tell the user "There's a problem so go find it" instead of just stating what the conflict is.

Reply:
definitely agree